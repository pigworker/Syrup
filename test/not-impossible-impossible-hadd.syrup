!<Bit> -> <Bit>
!X = Z where
  Z = nand(X, X)

<Bit> & <Bit> -> <Bit>
X & Y = Z where
  A = nand(X, Y)
  Z = !A

<Bit> | <Bit> -> <Bit>
X | Y = Z where
  A, B = !X, !Y
  Z = nand(A, B)

one() -> <Bit>
one() = !zero()

xor(<Bit>, <Bit>) -> <Bit>
xor(X, Y) = Z where
  A, B = X | Y, X & Y
  C = !B
  Z = A & C

hadd(<Bit>,<Bit>) -> <Bit>,<Bit>
fadd(<Bit>,<Bit>,<Bit>) -> <Bit>,<Bit>
tripleAddTwo(<Bit>,<Bit>) -> <Bit>,<Bit>,<Bit>,<Bit>


hadd(x,y) = a2, b1 where
  a2 = and(x,y)
  a1 = xor(x,y)

experiment hadd

fadd(Cin,Xin,Yin) = Cout,Zout where
  a2,a1 = hadd(Xin,Yin)
  b2,Zout = hadd(Cin,a1)
  Cout = xor(a2,b2)

experiment fadd
display fadd

tripleAddTwo