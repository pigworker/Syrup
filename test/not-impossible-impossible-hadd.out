Circuit not is defined.

Circuit and is defined.

Circuit or is defined.

Circuit one is defined.

Circuit xor is defined.

Error: I was trying to make sense of the following code:

hadd(<Bit>,<Bit>) -> <Bit>,<Bit>
hadd(x,y) = a2, b1 where
  a2 = and(x,y)
  a1 = xor(x,y)


Error: You tried to use b1 but it is not in scope.

Warning: Circuit hadd has been stubbed out.

Circuit fadd is defined.

Warning: the wire a1 is defined but never used.

Error: You haven't defined the circuit tripleAddTwo just now.

Warning: Circuit tripleAddTwo has been stubbed out.

Truth table for hadd:
    |    
----|----
0 0 | ? ?
0 1 | ? ?
1 0 | ? ?
1 1 | ? ?

Truth table for fadd:
Cin Xin Yin | Cout Zout
------------|----------
0   0   0   | ?    ?   
0   0   1   | ?    ?   
0   1   0   | ?    ?   
0   1   1   | ?    ?   
1   0   0   | ?    ?   
1   0   1   | ?    ?   
1   1   0   | ?    ?   
1   1   1   | ?    ?   

Warning: When displaying fadd:
  Warning: I don't have an implementation for hadd.
  Warning: I don't have an implementation for hadd.

Error: I was trying to make sense of the following code:

tripleAddTwo

I got stuck here: 
   {HERE} tripleAddTwo 
I was looking for Syrup source code.
At that point, I didn't know where to begin.
